# --------------------------------------------------------------------------------

 <h1 style="color:red;">  DATA 3550 Fall 2023 Homework 1 </h1>
 
 * (Due: Sunday, September 17, 2023 by 5 pm)
 
 * Following the formatting guidelines, please upload the .ipynb file in the D2L dropbox folder for Homework 1.

# --------------------------------------------------------------------------------

#  Purpose of Homework 1

- Basic Python uses
- Data Visualization
- Clean and increase the quality of a messy dataset by handling missing values, inconsistencies and anomalies in the dataset
- Create the dummy variables from the categorical variables
- Merge the datasets


## Information about the Dataset

## Dataset description

#### Major League Baseball Data from the 1986 and 1987 seasons.
** A data frame with 322 observations of major league players on the following 20 variables **

* AtBat: Number of times at bat in 1986
* Hits: Number of hits in 1986
* HmRun: Number of home runs in 1986
* Runs: Number of runs in 1986
* RBI: Number of runs batted in in 1986
* Walks: Number of walks in 1986
* Years: Number of years in the major leagues
* CAtBat: Number of times at bat during his career
* CHits: Number of hits during his career
* CHmRun: Number of home runs during his career
* CRuns: Number of runs during his career
* CRBI: Number of runs batted in during his career
* Walks: Number of walks during his career
* League: A factor with levels A and N indicating player's league at the end of 1986
* Division: A factor with levels E and W indicating player's division at the end of 1986
* PutOuts: Number of put outs in 1986
* Assists: Number of assists in 1986
* Errors: Number of errors in 1986
* Salary: 1987 annual salary on opening day in thousands of dollars
* NewLeague: A factor with levels A and N indicating player's league at the beginning of 1987


# Homework Submission Guidelines

## Formatting Guidelines

It is very important that your Jupyter Notebook is formatted correctly with markdown, comments, and code that works.
Please do the following for each section (See Example below):

* Include a title as a Heading 2
* Include a brief summary of the section
* Include your code and make sure it is executable and correct, include comments with the code.
* At the end of the section, include a brief summary of the results.

## How to turn it in:

*  Your Jupyter notebook file must be named HW02_LastnameFirstInitial.ipynb. For example: the file name will be HW02_RimalR.ipynb if I submit the Homework 2.
* You are to turn in your Jupyter notebook file only. No data files and no folders.
* It is assumed that you created your Jupyter notebook in a folder named HW02_student and in that folder is a Dataset folder. It is expected the path for importing a file is looking for a data folder, for example ‘Dataset/studentInfo.csv’.

## Example
* To answer the part a) and part b) of question 2,  please do the following.

* This section import the dataset Hitters.csv to the Jupyter notebook and explores the basic information of the dataset

## 2) Import the datasets

## a) Import the dataset Hitters.csv

* The following code import the Hitters.csv into the notebook and assign the name  df_Hi to the data

In [3]:
import pandas as pd 
df_Hi = pd.read_csv('Hitters.csv')

## b) Print the info of  df_Hi and explain

* The following code block provides a concise summary of the dataframe df_Hi, for an exploratory data analysis.

In [5]:
df_Hi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 20 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AtBat      322 non-null    int64  
 1   Hits       322 non-null    int64  
 2   HmRun      322 non-null    int64  
 3   Runs       322 non-null    int64  
 4   RBI        322 non-null    int64  
 5   Walks      322 non-null    int64  
 6   Years      322 non-null    int64  
 7   CAtBat     322 non-null    int64  
 8   CHits      322 non-null    int64  
 9   CHmRun     322 non-null    int64  
 10  CRuns      322 non-null    int64  
 11  CRBI       322 non-null    int64  
 12  CWalks     322 non-null    int64  
 13  League     322 non-null    object 
 14  Division   322 non-null    object 
 15  PutOuts    322 non-null    int64  
 16  Assists    322 non-null    int64  
 17  Errors     322 non-null    int64  
 18  Salary     263 non-null    float64
 19  NewLeague  322 non-null    object 
dtypes: float64

* We observed that the DataFrame df_Hi consists of 20 variables with 322 observations. Among them, 17 variables are numeric variables and 3 of them are categorical. The variable salary has missing values as the total number of non_null count is 263 which is less than the total number of obesrvations. The other variables has no missing values.

## 1. Import required packages

In [7]:

import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
import missingno as msno
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')


## 2. Import  the datasets
a) Import  the dataset Hitters.csv and assign it to df_Hi

b) Print the info of the pandas DataFrame df_Hi and explain your observations.



## 3. Data Visualization

a) Create a visualization to observe the missing values and their patterns in the df_Hi and explain your observations.

b) Create a countplot of the all three categorical variables in the data and explain your observations.

c) Create a distribution plot of all numerical variables. Breifly explain what you observe from those plots.


d) Create the comparative boxplot of all the numerical variables. Explain your observation.

## 4. Data Exploration

a) Calculate the statistical summary of the data df_Hi. Explain your observations.

b) Observe the unique number of values, most repeated values, and least repeated values of the variables.

c) Check if there is any outliers in the dataset.

## 5. Imputing Missing Values 

1) Check if there is any missing values in the Salary column of DataFrame df_Hi. What did you notice?

2)  How do we impute the missing values for df_Hi? Why did you use that, write your reasoning.

3) Verify that you correctly imputed the missing values.



## 6. Create the Dummy Variables

a) Create the dummies of the variable NewLeague. What is the count of each category?

b) Create the dummies of the variable League. What is the count of each category?

c) Create the dummies of the variable Division. What is the count of each category?

## 7. Merge the Data and Perform the Correlation Analysis

a) Merge the dummies created with the DataFrame df_Hi.

b) Look at the info of the data, how many variables are there now?

c) Create the heatmap that shows the pairwise correlation between all the variables.

d) Observe the correlation coefficients and identify the pairs that has a correlation of more than 0.8 in absolute value. 

e) Successively drop those variables until  there is no variable with the pairwise correlation higher than 0.8. 

## 8. Transform the data.

a) Transform the column Salary to the binary numerical variable as follows: If the Salary is above the median salary assign the value 1, otherwise assign the value 0. 


b) Verify that you correctly transformed the data. Observe the count of 1 and 0 in this column after the transformation.

## 9. Bonus Problem

a) Create any interesting data visualizations that may provide further insights of the data. You can do anything that make sense to you. Please explain your work properly.
    
b) Create new features that you think is important and explain why is it important?    